src/table/io/FitsSchemaInputMapper.cc

```diff
@@ -0,0 +1,619 @@
+    // on column number, flag bit, and name (ttype).  This allows us to insert fields into the
+    // schema in the correct order, regardless of which order they appear in the
+    // FITS header.
+    typedef boost::multi_index_container<
+        FitsSchemaItem,
+        boost::multi_index::indexed_by<
+            boost::multi_index::ordered_non_unique<
+                boost::multi_index::member<FitsSchemaItem,int,&FitsSchemaItem::column>
+            >,
+            boost::multi_index::ordered_non_unique<
+                boost::multi_index::member<FitsSchemaItem,int,&FitsSchemaItem::bit>
+            >,
+            boost::multi_index::hashed_unique<
+                boost::multi_index::member<FitsSchemaItem,std::string,&FitsSchemaItem::ttype>
+            >,
+            boost::multi_index::sequenced<>
+        >
+    > InputContainer;
+
+    // Typedefs for the special functors used to set data members.
+    typedef SetFitsSchemaString<&FitsSchemaItem::ttype> SetTTYPE;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tform> SetTFORM;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tccls> SetTCCLS;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tunit> SetTUNIT;
+    typedef SetFitsSchemaString<&FitsSchemaItem::doc> SetDoc;
+
+    // Typedefs for the different indices.
+    typedef InputContainer::nth_index<0>::type ByColumn;
+    typedef InputContainer::nth_index<1>::type ByBit;
+    typedef InputContainer::nth_index<2>::type ByName;
+    typedef InputContainer::nth_index<3>::type AsList;
+
+    // Getters for the different indices.
+    ByColumn & byColumn() { return inputs.get<0>(); }
+    ByBit & byBit() { return inputs.get<1>(); }
+    ByName & byName() { return inputs.get<2>(); }
+    AsList & asList() { return inputs.get<3>(); }
+
+    Impl() : version(0), flagColumn(0), archiveHdu(-1) {}
+
+    int version;
+    int flagColumn;
+    int archiveHdu;
+    Schema schema;
+    std::vector<std::unique_ptr<FitsColumnReader>> readers;
+    std::vector<Key<Flag>> flagKeys;
+    std::unique_ptr<bool[]> flagWorkspace;
+    PTR(io::InputArchive) archive;
+    InputContainer inputs;
+};
+
+FitsSchemaInputMapper::FitsSchemaInputMapper(daf::base::PropertyList & metadata, bool stripMetadata) :
+    _impl(boost::make_shared<Impl>())
+{
+    // Set the table version.  If AFW_TABLE_VERSION tag exists, use that
+    // If not, set to 0 if it has an AFW_TYPE, Schema default otherwise (DM-590)
+    if (!metadata.exists("AFW_TYPE")) {
+        _impl->version = lsst::afw::table::Schema::VERSION;
+    }
+    _impl->version = metadata.get("AFW_TABLE_VERSION", _impl->version);
+    if (stripMetadata) {
+        metadata.remove("AFW_TABLE_VERSION");
+    }
+
+    // Find a key that indicates an Archive stored on other HDUs
+    _impl->archiveHdu = metadata.get("AR_HDU", -1);
+    if (stripMetadata && _impl->archiveHdu > 0) {
+        metadata.remove("AR_HDU");
+    }
+
+    // read aliases stored in the new, expected way
```

Comment:
Should we define what the "new, expected way" is? And what if there are aliases stored in an old way (is there an old way)?

More generally, I'm a bit conflicted as to what extent the on-disk structure of the FITS files should be documented. Certainly, if we ever expected anybody to understand the FITS files without reading them through the LSST code, we need to document things like how aliases are stored. I'm not sure if that's necessary though; maybe it's legit to say that this is basically an opaque blob which you need to read with the LSST stack. Even in that latter case, it might also be convenient for our own purposes to document how we're storing things in the files.